drv/as5048b/as5048b.c

```diff
@@ -0,0 +1,76 @@
+/**
+ * @file
+ * @ingroup     drv_as5048b
+ * @author      Diego Badillo-San-Juan <diego.badillo-san-juan@inria.fr>
+ * @brief       Module for reading the AS5048B rotary encoder
+ *
+ * @copyright   Inria, 2024
+ *
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+
+// Include BSP packages
+#include "board_config.h"
+#include "gpio.h"
+#include "i2c.h"
+#include "as5048b.h"
+
+//=========================== defines ==========================================
+
+// Hardware settings of pins A1 and A2 define the last two bits of the slave address
+// Set A1 and A2 to GND to get default address 0x40
+#define AS5048B_SLAVE_ADDRESS (0x40)
+#define SLAVE_ADDRESS_REG     (0x15)
+
+// Registers where absolute angle is stored
+// 14 bits in total, 6-13(MSB) and 0-5(LSB) respectively
+#define AS5048B_ANGLE_8MSB (0xFE)
+#define AS5048B_ANGLE_6LSB (0xFF)
+
+//============================== public ========================================
+
+// Initialise I2C communication with rotary encoder
+void as5048b_init(void) {
+    // AS5048B doesn't have a WHO_AM_I register
+    db_i2c_init(&db_scl, &db_sda);
+}
+
+// Reads two 8-bit registers where the 14-bit raw absolute angle is stored
+uint16_t as5048b_i2c_read_raw_angle(void) {
+    uint8_t data_registers[2];
+
+    // Read registers directly via I2C
+    db_i2c_begin();
+    db_i2c_read_regs(AS5048B_SLAVE_ADDRESS, AS5048B_ANGLE_8MSB, &data_registers[0], 1);
+    db_i2c_read_regs(AS5048B_SLAVE_ADDRESS, AS5048B_ANGLE_6LSB, &data_registers[1], 1);
+    db_i2c_end();
+
+    // Combine MSB and LSB to get the 14-bit angle value
+    return (uint16_t)(data_registers[0] << 6) | data_registers[1];
+}
+
+// Read angle as a float in radians
+float as5048b_i2c_read_angle_radian(void) {
+    uint16_t angle_uint = as5048b_i2c_read_raw_angle();
+
+    // Convert 14-bit raw angle (0x0 to 0x3FFF) to radians [0, 2*M_PI)
+    return (float)as5048b_convert_raw_angle(angle_uint, 2 * M_PI);
+}
+
+// Read angle as a float in degrees
+float as5048b_i2c_read_angle_degree(void) {
+    uint16_t angle_uint = as5048b_i2c_read_raw_angle();
+
+    // Convert 14-bit raw angle (0x0 to 0x3FFF) to degrees [0, 360)
+    return (float)as5048b_convert_raw_angle(angle_uint, 360.);
+}
+
+// Convert the raw angle to [0, max_angle)
```

Comment:
```suggestion
// Convert the raw angle to [0, max_angle]
```
or the closing parenthesis has a specific meaning that I don't get ? ;)

Reply:
I meant that the max_angle is not considered. For example if we are converting the raw angle to float degrees, the max_angle = 360.0 is not considered, since that would be 0

Reply:
Ok, I didn't know (or remember...) that notation. I'm used to `[0, max_angle[` that is the same :)